### ISSUE TITLE
Cannot convert null to T?

### ISSUE THREAD

**Issue by MattConroy** [dotnet/roslyn#53139]
**Version Used**: .NET SDK 5.0.202

**Steps to Reproduce**:

1. Add the following code into any C# 9.0 project
```csharp
class Program
{
    static void Main(string[] args)
    {
    }

    public T? GetNull<T>()
    {
        return null;
    }
}
```
2. Build the project via `dotnet build`

**Expected Behavior**:
No errors are produced.

**Actual Behavior**:
The following error is produced:
`error CS0403: Cannot convert null to type parameter 'T' because it could be a non-nullable value type. Consider using 'default(T)' instead.`

I'm not sure if this is a bug I've discovered or a misunderstanding of how Nullable Reference Types are supposed to work. I was expecting to be able to return null from a method that describes that it returns T? rather than T. The above error is displayed regardless of if T is unconstrained, constrained to `notnull` or constrained to `class`.

**Comment by canton7** [comment]
> The above error is displayed regardless of if T is unconstrained, constrained to `notnull` or constrained to `class`.

Not true. `where T : class` [does remove the error](https://sharplab.io/#v2:EYLgxg9gTgpgtADwGwBYA0AXEBLANgHwGIA7AV11wENhcYACGY62gWACgABAJjo4EYA7OwDe7OuN58kvFHQCylbMQAU/AAwBtALp1KUAOYBnAJRiJothLoBfdmfEcAzHQAqAfjoBxGBgBy5XAAeFwA+ZWM6AHcACxhYVzoQXi57OgsrKw4BOjIKAG5U2zZrIA===), as does `where T : struct`.

The problem is that `T?` in an unconstrained generic method doesn't mean "If `T` is a value type, then `T?` means `Nullable<T>`" (it would be lovely if this was the case, but the runtime doesn't support it). `T?` in this context means "If `T` is a reference type then allow `null`, otherwise if `T` is a value type this means the same as `T`". Concerns around this were raised during the design process, but the design ultimately went ahead.

Contrast that to `T?` in a method where `T` is constrained to a value type, where `T?` *does* mean `Nullable<T>`. 

For example:

```cs
T? M<T>() => default;

var result = M<int>(); // 'result' has the type 'int', not 'int?'
```

I do also find the compiler error here confusing, though: it hasn't changed since before NRTs were introduced.

**Comment by MattConroy** [comment]
> The problem is that `T?` in an unconstrained generic method doesn't mean "If `T` is a value type, then `T?` means `Nullable<T>`" (it would be lovely if this was the case, but the runtime doesn't support it). `T?` in this context means "If `T` is a reference type then allow `null`, otherwise if `T` is a value type this means the same as `T`".

Aha! This is the part that tripped me up. Thanks!

I see how NRTs don't really affect this now, but it does make it confusing when trying to work with them regarding this issue. This may be a Visual Studio issue rather than a compiler one, but the underlined text here, is especially confusing.

![image](https://user-images.githubusercontent.com/9549279/117014212-7d545100-ace8-11eb-8d92-cbd50d50ebd0.png)
"Cannot convert expression type 'null' to return type 'T?'"

Though with your explanation that T? is not Nullable<T> for reference type it does make sense.
Would it be worth improving some of the compiler error messages around this?

**Comment by Youssef1313** [comment]
@MattConroy The underlined text doesn't seem to be coming from Roslyn. Could it be from ReSharper?

**Comment by MattConroy** [comment]
@Youssef1313 Looks like it absolutely is, or at least disabling ReSharper removes the underlined text.

The remaining message is clearer on its own, so I'll take it up with JetBrains!

Thanks guys 👍
